Tutorials/Heisenberg1d.ipynb

```diff
@@ -0,0 +1,718 @@
+    "ax1.legend()\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Well that's not too bad for a simple ansatz. But we can do better, can't we?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Learning with a Restricted Boltzmann Machine (RBM)\n",
+    "\n",
+    "We will now consider another celebrated ansatz, the Restricted Boltzmann Machine (RBM). It simply consists of two layers: a visible one representing the $L$ spin 1/2 degrees of freedom, and an hidden one which contains a different number $M$ of hidden units. There are connections between all visible and hidden nodes. The ansatz is the [following](https://www.netket.org/docs/machine_RbmSpin/)\n",
+    "\n",
+    "$\\Psi_{\\rm RBM} (\\sigma_1^z,\\sigma_2^z, ..., \\sigma_L^z)  = \\exp ( \\sum_{i=1}^L a_i \\sigma_i^z ) \\prod_{i=1}^M \\cosh (b_i + \\sum_j W_{ij} \\sigma^z_j)$ \n",
+    "\n",
+    "$a_i$ (resp. $b_i$) are the visible (resp. hidden) bias. Together with the weights $W_{ij}$, they are variational parameters that we (or rather NetKet) will optimize to minimize the energy. Netket gives you the control on the important parameters in this ansatz, such as $M$ and the fact that you want to use or not the biases. The full explanation is [here](https://www.netket.org/docs/machine_RbmSpin/). \n",
+    "\n",
+    "More conveniently (especially if you want to try another $L$ in this tutorial), let's define the hidden unit density $\\alpha = M / L$, and invoke the RBM ansatz in NetKet with as many hidden as visible units. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 131,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# RBM ansatz with alpha=1\n",
+    "ma = nk.machine.RbmSpin(alpha=1, hilbert=hi)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And let us use the same sampler (Metropolis exchange) with some different random initial parameters, optimizer (stochastic gradient), and variational method (stochastic reconfiguration) as for the Jastrow ansatz, and let's run things!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 132,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "### RBM calculation\n",
+      "Has 528 parameters\n",
+      "The RBM calculation took 79.31553912162781 seconds\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Sampler\n",
+    "sa = nk.sampler.MetropolisExchange(machine=ma,graph=g)\n",
+    "ma.init_random_parameters(seed=123, sigma=0.01)\n",
+    "\n",
+    "# Optimizer\n",
+    "op = nk.optimizer.Sgd(learning_rate=0.05)\n",
+    "# Stochastic reconfiguration\n",
+    "gs = nk.variational.Vmc(\n",
+    "    hamiltonian=ha,\n",
+    "    sampler=sa,\n",
+    "    optimizer=op,\n",
+    "    n_samples=1000,\n",
+    "    diag_shift=0.1,\n",
+    "    use_iterative=True,\n",
```

Comment:
Maybe it's even faster without the iterative solver on this small system? [not sure]

Reply:
No I think the iterative solver is still faster unless symmetries are used.